File: app/src/main/java/mega/privacy/android/app/SearchNodesTask.java
package mega.privacy.android.app;

import android.content.Context;
import android.os.AsyncTask;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;

import java.util.ArrayList;
import java.util.List;

import mega.privacy.android.app.lollipop.CloudDriveExplorerFragmentLollipop;
import mega.privacy.android.app.lollipop.FileExplorerActivityLollipop;
import mega.privacy.android.app.lollipop.IncomingSharesExplorerFragmentLollipop;
import mega.privacy.android.app.lollipop.ManagerActivityLollipop;
import mega.privacy.android.app.lollipop.managerSections.SearchFragmentLollipop;
import nz.mega.sdk.MegaApiAndroid;
import nz.mega.sdk.MegaApiJava;
import nz.mega.sdk.MegaCancelToken;
import nz.mega.sdk.MegaNode;
import nz.mega.sdk.MegaShare;

import static mega.privacy.android.app.fragments.managerFragments.LinksFragment.getLinksOrderCloud;
import static mega.privacy.android.app.lollipop.ManagerActivityLollipop.*;
import static mega.privacy.android.app.utils.LogUtil.*;
import static mega.privacy.android.app.utils.SortUtil.*;
import static mega.privacy.android.app.utils.TextUtil.isTextEmpty;
import static nz.mega.sdk.MegaApiJava.INVALID_HANDLE;

public class SearchNodesTask extends AsyncTask<Void, Void, Void> {
    private Context context;
    private ManagerActivityLollipop managerA;
    private FileExplorerActivityLollipop fileExplorerA;
    private IncomingSharesExplorerFragmentLollipop iSharesExplorerF;
    private CloudDriveExplorerFragmentLollipop cDExplorerF;
    private SearchFragmentLollipop searchF;

    private MegaApiAndroid megaApi;
    private MegaCancelToken megaCancelToken;

    private String query;
    private long parentHandleSearch;
    private ArrayList<MegaNode> nodes;

    private int orderCloud;
    private int orderOthers;

    public SearchNodesTask(Context mContext, Fragment mFragment, String mQuery, long mParentHandleSearch, ArrayList<MegaNode> mNodes){

        context = mContext;

        if (context instanceof ManagerActivityLollipop) {
            managerA = (ManagerActivityLollipop) context;
        } else if (context instanceof FileExplorerActivityLollipop) {
            fileExplorerA = (FileExplorerActivityLollipop) context;
        }

        if (mFragment instanceof SearchFragmentLollipop) {
            searchF = (SearchFragmentLollipop) mFragment;
        } else if (mFragment instanceof IncomingSharesExplorerFragmentLollipop) {
            iSharesExplorerF = (IncomingSharesExplorerFragmentLollipop) mFragment;
        } else if (mFragment instanceof CloudDriveExplorerFragmentLollipop) {
            cDExplorerF = (CloudDriveExplorerFragmentLollipop) mFragment;
        }

        query = mQuery;
        parentHandleSearch = mParentHandleSearch;
        nodes = mNodes;

        getOrder();

        megaApi = MegaApplication.getInstance().getMegaApi();
    }

    @Override
    protected Void doInBackground(Void... voids) {
        getSearchNodes();
        return null;
    }

    @Override
    protected void onPostExecute(Void aVoid) {
        if (isSearchF()) {
            searchF.setProgressView(false);
            searchF.setNodes(nodes);
        } else if (isCDExplorerF()) {
            cDExplorerF.setProgressView(false);
            cDExplorerF.setSearchNodes(nodes);
        } else if (isISharesExplorerF()) {
            iSharesExplorerF.setProgressView(false);
            iSharesExplorerF.setSearchNodes(nodes);
        }
    }

    public void cancelSearch() {
        if (megaCancelToken != null && !megaCancelToken.isCancelled()) {
            megaCancelToken.cancel();
        }
    }

    private void getOrder() {
        if (isSearchF()) {
            orderCloud = managerA.orderCloud;
            orderOthers = managerA.orderOthers;
        } else {
            MegaPreferences prefs = DatabaseHandler.getDbHandler(context).getPreferences();

            if (prefs != null && prefs.getPreferredSortCloud() != null) {
                orderCloud = Integer.parseInt(prefs.getPreferredSortCloud());
            } else {
                orderCloud = MegaApiJava.ORDER_DEFAULT_ASC;
            }

            if (prefs != null && prefs.getPreferredSortOthers() != null) {
                orderOthers = Integer.parseInt(prefs.getPreferredSortOthers());
            } else {
                orderOthers = MegaApiAndroid.ORDER_DEFAULT_ASC;
            }
        }
    }

    private void getSearchNodes() {
        if (query == null) {
            nodes.clear();
            return;
        }

        MegaNode parent = null;
        long parentHandle;

        if (parentHandleSearch == INVALID_HANDLE) {
            if (isSearchF()) {
                final ManagerActivityLollipop.DrawerItem drawerItem = managerA.getSearchDrawerItem();
                if (drawerItem == null) return;

                switch (drawerItem) {
                    case HOMEPAGE:
                        parent = megaApi.getRootNode();
                        break;
                    case CLOUD_DRIVE: {
                        parent = megaApi.getNodeByHandle(managerA.getParentHandleBrowser());
                        break;
                    }
                    case SHARED_ITEMS: {
                        if (managerA.getSearchSharedTab() == INCOMING_TAB) {
                            if (managerA.getParentHandleIncoming() == INVALID_HANDLE) {
                                getInShares();
                                return;
                            }

                            parent = megaApi.getNodeByHandle(managerA.getParentHandleIncoming());
                        } else if (managerA.getSearchSharedTab() == OUTGOING_TAB) {
                            if (managerA.getParentHandleOutgoing() == INVALID_HANDLE) {
                                getOutShares();
                                return;
                            }

                            parent = megaApi.getNodeByHandle(managerA.getParentHandleOutgoing());
                        } else if (managerA.getSearchSharedTab() == LINKS_TAB) {
                            if (managerA.getParentHandleLinks() == INVALID_HANDLE) {
                                getLinks();
                                return;
                            }

                            parent = megaApi.getNodeByHandle(managerA.getParentHandleLinks());
                        }
                        break;
                    }
                    case RUBBISH_BIN: {
                        parentHandle = managerA.getParentHandleRubbish();
                        if (parentHandle == INVALID_HANDLE) {
                            parent = megaApi.getRubbishNode();
                        } else {
                            parent = megaApi.getNodeByHandle(managerA.getParentHandleRubbish());
                        }
                        break;
                    }
                    case INBOX: {
                        parentHandle = managerA.getParentHandleInbox();
                        if (parentHandle == INVALID_HANDLE) {
                            parent = megaApi.getInboxNode();
                        } else {
                            parent = megaApi.getNodeByHandle(parentHandle);
                        }
                        break;
                    }
                }
            } else if (isCDExplorerF()) {
                parent = megaApi.getNodeByHandle(fileExplorerA.getParentHandleCloud());
            } else if (isISharesExplorerF()) {
                if (fileExplorerA.getParentHandleIncoming() == INVALID_HANDLE) {
                    getInShares();
                    return;
                }

                parent = megaApi.getNodeByHandle(fileExplorerA.getParentHandleIncoming());
            }
        } else {
            parent = megaApi.getNodeByHandle(parentHandleSearch);
        }

        if (parent != null) {
            if (isTextEmpty(query) || parentHandleSearch != INVALID_HANDLE) {
                nodes = megaApi.getChildren(parent);
            } else {
                megaCancelToken = MegaCancelToken.createInstance();
                nodes = megaApi.search(parent, query, megaCancelToken, true, orderCloud);
            }
        }
    }

    /**
     * Gets search result nodes of Incoming section, root navigation level.
     */
    private void getInShares() {
        if (isTextEmpty(query)) {
            nodes = megaApi.getInShares();

            if (orderOthers == MegaApiJava.ORDER_DEFAULT_DESC) {
                sortByMailDescending(nodes);
            }
        } else {
            megaCancelToken = MegaCancelToken.createInstance();
            nodes = megaApi.searchOnInShares(query, megaCancelToken, orderCloud);
        }
    }

    /**
     * Gets search result nodes of Outgoing section, root navigation level.
     */
    private void getOutShares() {
        if (isTextEmpty(query)) {
            nodes.clear();

            ArrayList<MegaShare> outShares = megaApi.getOutShares();
            List<Long> addedHandles = new ArrayList<>();

            for (MegaShare outShare : outShares) {
                MegaNode node = megaApi.getNodeByHandle(outShare.getNodeHandle());

                if (node != null && !addedHandles.contains(node.getHandle())) {
                    addedHandles.add(node.getHandle());
                    nodes.add(node);
                }
            }

            if (orderOthers == MegaApiJava.ORDER_DEFAULT_DESC) {
                sortByNameDescending(nodes);
            } else {
                sortByNameAscending(nodes);
            }
        } else {
            megaCancelToken = MegaCancelToken.createInstance();
            nodes = megaApi.searchOnOutShares(query, megaCancelToken, orderCloud);
        }
    }

    /**
     * Gets search result nodes of Links section, root navigation level.
     */
    private void getLinks() {
        if (isTextEmpty(query)) {
            nodes = megaApi.getPublicLinks(getLinksOrderCloud(managerA.orderCloud, managerA.isFirstNavigationLevel()));
        } else {
            megaCancelToken = MegaCancelToken.createInstance();
            nodes = megaApi.searchOnPublicLinks(query, megaCancelToken, orderCloud);
        }
    }

    private boolean isSearchF() {
        return managerA != null && searchF != null;
    }

    private boolean isFileExplorerA() {
        return fileExplorerA != null;
    }

    private boolean isISharesExplorerF() {
        return isFileExplorerA() && iSharesExplorerF != null;
    }

    private boolean isCDExplorerF() {
        return isFileExplorerA() && cDExplorerF != null;
    }

    public static void setSearchProgressView(RelativeLayout contentLayout, ProgressBar searchProgressBar, RecyclerView recyclerView, boolean inProgress) {
        if (contentLayout == null || searchProgressBar == null || recyclerView == null) {
            logWarning("Cannot set search progress view, one or more parameters are NULL.");
            logDebug("contentLayout: " + contentLayout + ", searchProgressBar: " + searchProgressBar + ", recyclerView: " + recyclerView);
            return;
        }

        contentLayout.setEnabled(!inProgress);
        if (inProgress) {
            contentLayout.setAlpha(0.4f);
            searchProgressBar.setVisibility(View.VISIBLE);
            recyclerView.setVisibility(View.GONE);
        } else {
            contentLayout.setAlpha(1);
            searchProgressBar.setVisibility(View.GONE);
            recyclerView.setVisibility(View.VISIBLE);
        }
    }

    public static ArrayList<MegaNode> getSearchedNodes(ArrayList<String> handles) {
        long[] handleValues = new long[handles.size()];

        for (int i = 0; i < handles.size(); i++) {
            handleValues[i] = Long.parseLong(handles.get(i));
        }

        return getSearchedNodes(handleValues);
    }

    public static ArrayList<MegaNode> getSearchedNodes(long[] handles) {
        MegaApiAndroid megaApi = MegaApplication.getInstance().getMegaApi();
        ArrayList<MegaNode> nodes = new ArrayList<>();

        if (handles != null) {
            for (long handle : handles) {
                MegaNode node = megaApi.getNodeByHandle(handle);
                if (node != null) {
                    nodes.add(node);
                }
            }
        }

        return nodes;
    }
}

File: app/src/main/java/mega/privacy/android/app/search/SearchNodesTask.kt
package mega.privacy.android.app.search

import android.os.AsyncTask
import android.view.View
import android.widget.ProgressBar
import android.widget.RelativeLayout
import androidx.recyclerview.widget.RecyclerView
import mega.privacy.android.app.MegaApplication
import mega.privacy.android.app.di.MegaApi
import mega.privacy.android.app.fragments.managerFragments.LinksFragment
import mega.privacy.android.app.globalmanagement.SortOrderManagement
import mega.privacy.android.app.lollipop.ManagerActivityLollipop.*
import mega.privacy.android.app.utils.Constants.INVALID_VALUE
import mega.privacy.android.app.utils.LogUtil.logDebug
import mega.privacy.android.app.utils.LogUtil.logWarning
import mega.privacy.android.app.utils.SortUtil
import mega.privacy.android.app.utils.TextUtil
import nz.mega.sdk.*
import nz.mega.sdk.MegaApiJava.INVALID_HANDLE
import nz.mega.sdk.MegaApiJava.ORDER_DEFAULT_DESC
import java.util.ArrayList
import javax.inject.Inject

/**
 * Class which manages search actions.
 *
 * @property megaApi             MegaApiAndroid object.
 * @property sortOrderManagement SortOrderManagement object.
 * @property query               Typed text for search.
 * @property parentHandleSearch  INVALID_HANDLE if has to get the parent handle to search yet,
 *                               a valid handle if not.
 * @property parentHandle        Parent handle where the search has to be made.
 * @property nodes               List of MegaNodes shown before make the search.
 * @property callback            Callback to perform final actions.
 * @property searchType          It can be TYPE_GENERAL, TYPE_CLOUD_EXPLORER or TYPE_INCOMING_EXPLORER.
 */
class SearchNodesTask @Inject constructor(
    @MegaApi private val megaApi: MegaApiAndroid,
    private val sortOrderManagement: SortOrderManagement,
    private val query: String?,
    private val parentHandleSearch: Long,
    private val parentHandle: Long,
    private val nodes: ArrayList<MegaNode>,
    private val callback: Callback,
    private val searchType: Int
) : AsyncTask<Void, Void, Void>() {

    private var drawerItem: DrawerItem? = null
    private var sharesTab = INVALID_VALUE
    private var isFirstNavigationLevel = true

    private var megaCancelToken: MegaCancelToken? = null
    private var searchNodes = nodes

    /**
     * Constructor which manages TYPE_GENERAL search.
     *
     * @property megaApi                MegaApiAndroid object.
     * @property sortOrderManagement    SortOrderManagement object.
     * @property query                  Typed text for search.
     * @property parentHandleSearch     INVALID_HANDLE if has to get the parent handle to search yet,
     *                                  a valid handle if not.
     * @property parentHandle           Parent handle where the search has to be made.
     * @property nodes               List of MegaNodes shown before make the search.
     * @property callback               Callback to perform final actions.
     * @property searchType             It can be TYPE_GENERAL, TYPE_CLOUD_EXPLORER or TYPE_INCOMING_EXPLORER.
     * @property drawerItem             DrawerItem in which the search has to be made.
     * @property sharedTab              Tab of the Shares section if needed, INVALID_VALUE otherwise.
     * @property isFirstNavigationLevel True if is first navigation level, false otherwise.
     */
    constructor(
        @MegaApi megaApi: MegaApiAndroid,
        sortOrderManagement: SortOrderManagement,
        query: String?,
        parentHandleSearch: Long,
        parentHandle: Long,
        nodes: ArrayList<MegaNode>,
        callback: Callback,
        searchType: Int,
        drawerItem: DrawerItem,
        sharedTab: Int,
        isFirstNavigationLevel: Boolean
    ) : this(
        megaApi,
        sortOrderManagement,
        query,
        parentHandleSearch,
        parentHandle,
        nodes,
        callback,
        searchType
    ) {
        this.drawerItem = drawerItem
        this.sharesTab = sharedTab
        this.isFirstNavigationLevel = isFirstNavigationLevel
    }

    companion object {
        const val TYPE_GENERAL = 1
        const val TYPE_CLOUD_EXPLORER = 2
        const val TYPE_INCOMING_EXPLORER = 3

        /**
         * Updates the progress search view.
         *
         * @param contentLayout     General content view.
         * @param searchProgressBar ProgressBar to show or hide.
         * @param recyclerView      Container of the ProgressBar.
         * @param inProgress        True if the search is in progress, false otherwise.
         */
        @JvmStatic
        fun setSearchProgressView(
            contentLayout: RelativeLayout?,
            searchProgressBar: ProgressBar?,
            recyclerView: RecyclerView?,
            inProgress: Boolean
        ) {
            if (contentLayout == null || searchProgressBar == null || recyclerView == null) {
                logWarning("Cannot set search progress view, one or more parameters are NULL.")
                logDebug("contentLayout: $contentLayout, searchProgressBar: $searchProgressBar, recyclerView: $recyclerView")
                return
            }

            contentLayout.isEnabled = !inProgress

            if (inProgress) {
                contentLayout.alpha = 0.4f
                searchProgressBar.visibility = View.VISIBLE
                recyclerView.visibility = View.GONE
            } else {
                contentLayout.alpha = 1f
                searchProgressBar.visibility = View.GONE
                recyclerView.visibility = View.VISIBLE
            }
        }
    }

    override fun doInBackground(vararg voids: Void?): Void? {
        getSearchNodes()
        return null
    }

    override fun onPostExecute(aVoid: Void?) {
        callback.finishSearchNodes(searchNodes)
    }

    /**
     * Cancels the current search.
     */
    fun cancelSearch() {
        megaCancelToken?.cancel()
    }

    /**
     * Gets the searched nodes depending on the query typed and current section.
     */
    private fun getSearchNodes() {
        if (query == null) {
            searchNodes.clear()
            return
        }

        var parent: MegaNode? = null

        if (parentHandleSearch == INVALID_HANDLE) {
            when (searchType) {
                TYPE_GENERAL -> {
                    when (drawerItem) {
                        DrawerItem.HOMEPAGE -> parent = megaApi.rootNode
                        DrawerItem.CLOUD_DRIVE -> parent = megaApi.getNodeByHandle(parentHandle)
                        DrawerItem.SHARED_ITEMS -> {
                            when (sharesTab) {
                                INCOMING_TAB -> {
                                    if (parentHandle == INVALID_HANDLE) {
                                        getInShares()
                                        return
                                    }

                                    parent = megaApi.getNodeByHandle(parentHandle)
                                }
                                OUTGOING_TAB -> {
                                    if (parentHandle == INVALID_HANDLE) {
                                        getOutShares()
                                        return
                                    }

                                    parent = megaApi.getNodeByHandle(parentHandle)
                                }
                                LINKS_TAB -> {
                                    if (parentHandle == INVALID_HANDLE) {
                                        getLinks()
                                        return
                                    }

                                    parent = megaApi.getNodeByHandle(parentHandle)
                                }
                            }
                        }
                        DrawerItem.RUBBISH_BIN -> {
                            parent = if (parentHandle == INVALID_HANDLE) {
                                megaApi.rubbishNode
                            } else {
                                megaApi.getNodeByHandle(parentHandle)
                            }
                        }
                        DrawerItem.INBOX -> {
                            parent = if (parentHandle == INVALID_HANDLE) {
                                megaApi.inboxNode
                            } else {
                                megaApi.getNodeByHandle(parentHandle)
                            }
                        }
                        else -> parent = megaApi.rootNode
                    }
                }
                TYPE_CLOUD_EXPLORER -> {
                    parent = megaApi.getNodeByHandle(parentHandle)
                }
                TYPE_INCOMING_EXPLORER -> {
                    if (parentHandle == INVALID_HANDLE) {
                        getInShares()
                        return
                    }

                    parent = megaApi.getNodeByHandle(parentHandle)
                }
            }
        } else {
            parent = megaApi.getNodeByHandle(parentHandleSearch)
        }

        if (parent != null) {
            if (TextUtil.isTextEmpty(query) || parentHandleSearch != INVALID_HANDLE) {
                searchNodes = megaApi.getChildren(parent)
            } else {
                megaCancelToken = MegaCancelToken.createInstance()
                searchNodes = megaApi.search(
                    parent,
                    query,
                    megaCancelToken,
                    true,
                    sortOrderManagement.getOrderCloud()
                )
            }
        }
    }

    /**
     * Gets search result nodes of Incoming section, root navigation level.
     */
    private fun getInShares() {
        if (TextUtil.isTextEmpty(query)) {
            searchNodes = megaApi.getInShares(sortOrderManagement.getOrderOthers())
        } else {
            megaCancelToken = MegaCancelToken.createInstance()
            searchNodes = megaApi.searchOnInShares(
                query,
                megaCancelToken,
                sortOrderManagement.getOrderCloud()
            )
        }
    }

    /**
     * Gets search result nodes of Outgoing section, root navigation level.
     */
    private fun getOutShares() {
        if (TextUtil.isTextEmpty(query)) {
            searchNodes.clear()
            val outShares = megaApi.outShares
            val addedHandles: MutableList<Long> = ArrayList()
            for (outShare in outShares) {
                val node = megaApi.getNodeByHandle(outShare.nodeHandle)
                if (node != null && !addedHandles.contains(node.handle)) {
                    addedHandles.add(node.handle)
                    searchNodes.add(node)
                }
            }
            if (sortOrderManagement.getOrderOthers() == ORDER_DEFAULT_DESC) {
                SortUtil.sortByNameDescending(searchNodes)
            } else {
                SortUtil.sortByNameAscending(searchNodes)
            }
        } else {
            megaCancelToken = MegaCancelToken.createInstance()
            searchNodes = megaApi.searchOnOutShares(
                query,
                megaCancelToken,
                sortOrderManagement.getOrderCloud()
            )
        }
    }

    /**
     * Gets search result nodes of Links section, root navigation level.
     */
    private fun getLinks() {
        if (TextUtil.isTextEmpty(query)) {
            searchNodes = megaApi.getPublicLinks(
                LinksFragment.getLinksOrderCloud(
                    sortOrderManagement.getOrderCloud(), isFirstNavigationLevel
                )
            )
        } else {
            megaCancelToken = MegaCancelToken.createInstance()
            searchNodes = megaApi.searchOnPublicLinks(
                query,
                megaCancelToken,
                sortOrderManagement.getOrderCloud()
            )
        }
    }

    interface Callback {
        /**
         * Finishes the search.
         *
         * @param nodes List of the searched nodes.
         */
        fun finishSearchNodes(nodes: ArrayList<MegaNode>)
    }
}